jsk_2016_01_baxter_apc/euslisp/in-hand-data-collection-main.l

```diff
@@ -0,0 +1,258 @@
+                                       (arm-to-str arm_)) req)
+            :success))
+        (unix::sleep 1)
+        )
+      )
+    )
+  (:view-hand-pose
+    ()
+    (send *ri* :angle-vector
+          (send *baxter* :fold-to-keep-object arm_)
+          3000)
+    (send *ri* :wait-interpolation)
+    (send *ri* :angle-vector
+          (send *baxter* :view-hand-pose arm_ (elt bin-list_ (random (length bin-list_))))
+          3000)
+    (send *ri* :wait-interpolation)
+    (unix::sleep 1)
+    )
+  (:move-arm-to-view-table
+    ()
+    (if (eq arm_ :larm)
+      (send *baxter* :larm :angle-vector
+            #f(9.25049 42.3633 -132.869 125.244 140.471 107.71 -19.1602 0.0))
+      (send *baxter* :rarm :angle-vector
+            #f(-9.25049 42.3633 132.869 125.244 -140.471 107.71 19.1602 0.0))
+      )
+    (send *ri* :angle-vector (send *baxter* :angle-vector) 3000)
+    (send *ri* :wait-interpolation)
+    )
+  (:detect-object-on-table
+    ()
+    (send self :move-arm-to-view-table)
+    (let* ((msg) (frame_id) (pose))
+      (setq msg (one-shot-subscribe
+                  (format nil
+                          "cpi_decomposer_~a_table/centroid_pose_array"
+                          (arm-to-str arm_))
+                  geometry_msgs::PoseArray
+                  :after-stamp (ros::time-now)))
+      (if (or (null msg) (null (send msg :poses)))
+        (progn
+          (ros::ros-warn "[~a] No object is found on table. Retrying.." (ros::get-name))
+          (send self :detect-object-on-table)
+          )
+        (progn
+          (setq frame_id (send msg :header :frame_id))
+          (setq pose (elt (send msg :poses) 0))
+          (make-coords
+            :pos (send (send *ri* :tf-pose->coords frame_id pose) :worldpos))))
+      ))
+  (:pick-object-from-table
+    ()
+    (ros::ros-info "[~a] Picking object from table" (ros::get-name))
+    (let ((av0) (av1) (centroid (send self :detect-object-on-table)))
+      (send *ri* :gripper-servo-on arm_)
+      (setq av0 (send *baxter* arm_
+                     :inverse-kinematics
+                     (send (send centroid :copy-worldcoords)
+                           :translate (float-vector 0 0 230) :world)
+                     :use-gripper t
+                     :rotation-axis nil))
+      (unless av0 (return-from :pick-object-from-table))
+      (send *ri* :angle-vector av0 3000)
+      (send *ri* :wait-interpolation)
+      (setq av1
+            (send *baxter* arm_ :inverse-kinematics
+                  (send (send centroid :copy-worldcoords)
+                        :translate (float-vector 0 0 -50) :world)
+                  :use-gripper t
+                  :rotation-axis :z))
+      (unless av1 (return-from :pick-object-from-table))
+      (send *ri* :angle-vector av1 3000)
+      (send *ri* :wait-interpolation)
+      (send *ri* :start-grasp arm_)
+      (send *ri* :gripper-servo-off arm_)
+      (unix::sleep 3)
```

Comment:
Maybe `(unix::sleep 3)` -> `(send *ri* :gripper-servo-off arm_)` is good order because gripper may move before suction is stable when the order is `(send *ri* :gripper-servo-off arm_)` -> `(unix::sleep 3)`